chain/deals/request_validation.go

```diff
@@ -0,0 +1,229 @@
+package deals
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"reflect"
+
+	"github.com/ipfs/go-cid"
+	datastore "github.com/ipfs/go-datastore"
+	"github.com/ipfs/go-datastore/namespace"
+	"github.com/ipld/go-ipld-prime"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"go.uber.org/fx"
+
+	"github.com/filecoin-project/lotus/api"
+	"github.com/filecoin-project/lotus/datatransfer"
+	"github.com/filecoin-project/lotus/node/modules/dtypes"
+)
+
+var (
+	// ErrWrongVoucherType means the voucher was not the correct type can validate against
+	ErrWrongVoucherType = errors.New("cannot validate voucher type")
+
+	// ErrNoPushAccepted just means clients do not accept pushes for storage deals
+	ErrNoPushAccepted = errors.New("Client should not receive data for a storage deal")
+
+	// ErrNoPullAccepted just means providers do not accept pulls for storage deals
+	ErrNoPullAccepted = errors.New("Provider should not send data for a storage deal")
+
+	// ErrNoDeal means no active deal was found for this vouchers proposal cid
+	ErrNoDeal = errors.New("No deal found for this proposal")
+
+	// ErrWrongPeer means that the other peer for this data transfer request does not match
+	// the other peer for the deal
+	ErrWrongPeer = errors.New("Data Transfer peer id and Deal peer id do not match")
+
+	// ErrWrongPiece means that the pieceref for this data transfer request does not match
+	// the one specified in the deal
+	ErrWrongPiece = errors.New("Base CID for deal does not match CID for piece")
+
+	// ErrInacceptableDealState means the deal for this transfer is not in a deal state
+	// where transfer can be performed
+	ErrInacceptableDealState = errors.New("Deal is not a in a state where deals are accepted")
+
+	// AcceptableDealStates are the states in which it would make sense to actually start a data transfer
+	AcceptableDealStates = []api.DealState{api.DealAccepted, api.DealUnknown}
+)
+
+// StorageDataTransferVoucher is the voucher type for data transfers
+// used by the storage market
+type StorageDataTransferVoucher struct {
+	Proposal cid.Cid
+}
+
+// ToBytes converts the StorageDataTransferVoucher to raw bytes
+func (dv StorageDataTransferVoucher) ToBytes() []byte {
```

Comment:
I'm slightly inclined to use cbor-gen for this - it generates (Un)MarshalCBOR functions for you.

To use it:
* add the structs to `gen/main.go`
* `go run ./gen`
* gofmt
* fix imports (manually, just change it back to whatever was there before running gen) (it's a cbor-gen bug)

Reply:
Hey. I'd prefer not to require cbor-gen in the voucher interface, but this can still be CBOR-gen'd by calling the generated methods inside of FromBytes/ToBytes. That's what I will do and you can tell me if it works for you.